File: scrimage-core/src/main/scala/com/sksamuel/scrimage/canvas/Canvas.scala
package com.sksamuel.scrimage.canvas

import com.sksamuel.scrimage.{Color, Image}
import java.awt.{BasicStroke, Graphics2D, RenderingHints}
import scala.language.implicitConversions

case class Canvas(image: Image) {

  private def g2(image: Image): Graphics2D = image.awt().getGraphics.asInstanceOf[Graphics2D]

  protected[scrimage] def drawInPlace(drawable: Drawable): Unit = drawInPlace(Seq(drawable))
  protected[scrimage] def drawInPlace(drawables: Drawable*): Unit = drawInPlace(drawables)
  protected[scrimage] def drawInPlace(drawables: Iterable[Drawable]): Unit = {
    val g = g2(image)
    drawables.foreach { d =>
      d.context.configure(g)
      d.draw(g)
    }
    g.dispose()
  }

  def draw(drawables: Drawable*): Canvas = draw(drawables)
  def draw(drawables: Iterable[Drawable]): Canvas = {
    val target = image.copy
    val g = g2(target)
    drawables.foreach { d =>
      d.context.configure(g)
      d.draw(g)
    }
    g.dispose()
    new Canvas(target)
  }
}

object Canvas {

  implicit def image2canvas(image: Image): Canvas = new Canvas(image)
  implicit def canvas2image(canvas: Canvas): Image = canvas.image

  implicit class RichGraphics2(g2: Graphics2D) {

    def setBasicStroke(width: Float): Unit = g2.setStroke(new BasicStroke(width))

    def setWhite(): Unit = g2.setColor(Color.White.toAWT)
    def setBlack(): Unit = g2.setColor(Color.Black.toAWT)

    def setAntiAlias(aa: Boolean): Unit = {
      if (aa) {
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
        g2.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON)
      } else {
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_OFF)
        g2.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_OFF)
      }
    }
  }

}

File: scrimage-core/src/main/java/com/sksamuel/scrimage/canvas/Canvas.java
package com.sksamuel.scrimage.canvas;

import com.sksamuel.scrimage.ImmutableImage;
import com.sksamuel.scrimage.graphics.RichGraphics2D;

import java.awt.*;
import java.util.Arrays;
import java.util.Collection;

/**
 * A Canvas extends an ImmutableImage with functionality for drawing.
 * It is essentially some sugar around using the Graphics2D functionality in AWT.
 */
public class Canvas {

   private final ImmutableImage image;

   public Canvas(ImmutableImage image) {
      this.image = image;
   }

   /**
    * Returns the image that backs this canvas.
    *
    * @return the ImmutableImage that backs this canvas
    */
   public ImmutableImage getImage() {
      return image;
   }

   private Graphics2D g2(ImmutableImage image) {
      return (Graphics2D) image.awt().getGraphics();
   }

   public void drawInPlace(Drawable... drawables) {
      drawInPlace(Arrays.asList(drawables));
   }

   public void drawInPlace(Collection<Drawable> drawables) {
      Graphics2D g = g2(image);
      try {
         drawables.forEach(d -> drawScoped(g, d));
      } finally {
         g.dispose();
      }
   }

   public Canvas draw(Drawable... drawables) {
      return draw(Arrays.asList(drawables));
   }

   public Canvas draw(Collection<Drawable> drawables) {
      ImmutableImage target = image.copy();
      Graphics2D g = g2(target);
      try {
         drawables.forEach(d -> drawScoped(g, d));
      } finally {
         g.dispose();
      }
      return new Canvas(target);
   }

   // Each drawable gets a scratch Graphics2D snapshot so its GraphicsContext's
   // configure() mutations (setColor, setStroke, setFont, translate, rotate,
   // setComposite, …) don't bleed across to the next drawable. Without this
   // every subsequent drawable inherited whatever state the previous one
   // happened to leave on the shared Graphics2D — a red rectangle followed
   // by an un-configured one would render the second in red too.
   private static void drawScoped(Graphics2D g, Drawable d) {
      Graphics2D scoped = (Graphics2D) g.create();
      try {
         RichGraphics2D rich2d = new RichGraphics2D(scoped);
         d.context().configure(rich2d);
         d.draw(rich2d);
      } finally {
         scoped.dispose();
      }
   }
}
